src/System.Private.CoreLib/src/System/Exception.CoreCLR.cs

```diff
@@ -356,6 +354,8 @@ internal void RestoreDispatchState(in DispatchState dispatchState)
         // unless a System.Resolver object roots it.
         private readonly object? _dynamicMethods;
         private string? _source;         // Mainly used by VB.
+        // This is the object against which a lock will be taken when attempting to restore the EDI.
```

Comment:
Does this mean we're paying for this lock on exceptions that are never thrown with EDI? Or is the comment wrong?

Reply:
Managed code uses it for EDI; usmanaged code also uses it for the stacktrace (rather than the global spinlock it was using before)

Reply:
The comment should be updated then. Thanks.

Reply:
Hmm, did you see the contention in the spinlock on the VM side to be a problem? The global spin lock on the VM side is protecting a very small operation. It sounds unlikely that it would be a bottleneck during exception. There is a number of other similar global locks on the VM side that are taken during exception handling.

And if the tiny spin lock on the VM side was not a problem, then the @stephentoub's point about this not being pay-for-play holds.

Reply:
`SaveStackTracesFromDeepCopy` additionally grabs the spinlock so I wanted a reentrant lock which Monitor on object provides as it hits the spinlock in this path (as well as the regular exception path both for get and set).

@NickCraver likes to throw "a few million exceptions a minute" https://twitter.com/Nick_Craver/status/1177990124854005761

> You'd be amazed just how many exceptions per second we can throw while the site shows no symptoms on the surface.

I assume async exceptions feature heavily so they would also involve cascading EDI restores. While being able to throw exceptions even faster probably isn't helpful, dropping out the CPU spin contention likely is by having the lock local to each exception rather than a single global lock.

> There is a number of other similar global locks on the VM side that are taken during exception handling.

Hmm, I thought I got them all?

>  this not being pay-for-play holds.

The 480kB per 20k Exceptions for the lock; isn't that large in the scheme of throwing those 20k Exceptions which takes 346MB (which I'm trying to address in my other PRs)

![image](https://user-images.githubusercontent.com/1142958/65824858-8ad16b80-e267-11e9-9ea4-ec7e3efaa921.png)

Reply:
How much faster can you make throw exceptions after this change? This needs a performance tests that shows the improvement in number of exceptions thrown per second.

> Hmm, I thought I got them all?

There are locks of similar nature in stackwalking, etc.

Reply:
> dropping out the CPU spin contention

The operation protected by the spin lock was very tiny. I believe that the chances for that spin lock to spin are near zero.

Reply:
> This needs a performance tests that shows the improvement in number of exceptions thrown per second.

In parallel; is unlikely to change single thread exceptions. Will look to come up with something.

I've have issues with mass disconnect exceptions; though they are infrequent (e.g. app shutdown), and there are obviously other factors at play (socket shutdown etc)

Reply:
>  I believe that the chances for that spin lock to spin are near zero.

EDI.Capture, EDI.Throw, throw, ToString, getStacktrace etc all hit the spinlock; so there would be some cache contention. However, in the scheme of throwing exceptions, its likely very mild.

Approaching it from a different angle; for async as the stack unwinds this lock is hit again and again at every await. So throwing an exception in an async path doesn't grab this lock once, but again and again until it reaches the top level catch in user code.

TaskAwaiter.ThrowForNonSuccess, ExceutionContext.Run etc, follow this path
```
> EDI.Throw()
-> Exception.RestoreDispatchState
--> ** lock ** (in Exception.cs)
---> ExceptionNative::SaveStackTracesFromDeepCopy
----->ExceptionObject::SetStackTrace
-----> ** spin lock ** (in object.cpp)
-> throw
```
Given this do you feel it is worth moving the EDI lock away from a global lock?

I originally had it so the lock was lazily assigned via `Interlocked.CompareExchange` https://github.com/dotnet/coreclr/pull/26823#discussion_r326928194; but then changed it to always allocate it. Since it was now always allocated it seemed worthwhile to move the SpinLock to also use the same lock since it was there and also reentrant so it didn't matter if it was grabbed twice for the EDI.Throw path.

If it is worth moving the EDI lock away from the global lock; what alternative approach would you suggest?

Reply:
> Given this do you feel it is worth moving the EDI lock away from a global lock?

I have done a few tests that do EDI chaining on multiple threads and I do not see neither of the locks to be a problem. I think we should revert the change. The refactorings to reduce code duplication and deleting of the unnecessary code can stay.